### ISSUE TITLE
The training process cannot be reproduced with torch==2.1.2

### ISSUE THREAD

**Issue by AGI-player** [pytorch/pytorch#118089]
### 🐛 Describe the bug

The training process cannot be reproduced with torch==2.1.2, while it can be reprocduced by just replacing torch version 2.1.2 with 2.1.0

### Versions

with torch==2.1.0, the results are the same
<img width="385" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/540dbe96-f6a7-4ac1-83bf-9c4b8a68e7c1">
<img width="396" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/7a9e9b8a-e7c4-43df-be11-467c78c4a446">

with torch==2.1.2, the results are different
<img width="381" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/f3e06bf4-26f1-4ddd-b428-1da42d9cec73">
<img width="382" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/76cea51a-dfb1-4642-aed4-48bc69ced211">

**Comment by drisspg** [comment]
Hey would you mind providing more detail here, as well as a minimal reproducing script.

**Comment by AGI-player** [comment]
> Hey would you mind providing more detail here, as well as a minimal reproducing script.

The training code is based on https://github.com/hiyouga/LLaMA-Factory , run script as follows:
deepspeed
--num_gpus
8
/load/LLaMA-Factory/src/train_bash.py
--deepspeed
/load/LLaMA-Factory/src/ds_config_zero3.json
--stage
pt
--finetuning_type
full
--model_name_or_path
path to your model checkpoint
--output_dir
/save/ckpt/1
--overwrite_cache
--do_train
--per_device_train_batch_size
1
--gradient_accumulation_steps
2
--num_train_epochs
1
--lr_scheduler_type
cosine
--warmup_ratio
0.05
--learning_rate
1.5e-4
--adam_beta1
0.9
--adam_beta2
0.95
--adam_epsilon
1e-8
--weight_decay
0.1
--cutoff_len
4096
--bf16
True
--logging_steps
1
--save_steps
50000
--save_total_limit
3
--save_on_each_node
True
--overwrite_output_dir
--cache_path
path to your data(preprocessed)

ds_config_zero3.json as follow:
{
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu" :"auto",
    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "fp16": {
    "enabled": "auto",
    "loss_scale": 0,
    "loss_scale_window": 1000,
    "initial_scale_power": 16,
    "hysteresis": 2,
    "min_loss_scale": 1
    },
    "bf16": {
        "enabled": "auto"
    },
    "zero_optimization": {
    "stage": 3,
    "overlap_comm": true,
    "stage3_gather_16bit_weights_on_model_save": true
    }
}

conda environment as follow:
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             5.1                       1_gnu  
accelerate                0.26.1                   pypi_0    pypi
aiofiles                  23.2.1                   pypi_0    pypi
aiohttp                   3.9.1                    pypi_0    pypi
aiosignal                 1.3.1                    pypi_0    pypi
altair                    5.2.0                    pypi_0    pypi
annotated-types           0.6.0                    pypi_0    pypi
anyio                     4.2.0                    pypi_0    pypi
appdirs                   1.4.4                    pypi_0    pypi
async-timeout             4.0.3                    pypi_0    pypi
attrs                     23.2.0                   pypi_0    pypi
bzip2                     1.0.8                h7b6447c_0  
ca-certificates           2023.12.12           h06a4308_0  
certifi                   2023.11.17               pypi_0    pypi
charset-normalizer        3.3.2                    pypi_0    pypi
click                     8.1.7                    pypi_0    pypi
contourpy                 1.2.0                    pypi_0    pypi
cycler                    0.12.1                   pypi_0    pypi
datasets                  2.16.1                   pypi_0    pypi
deepspeed                 0.12.6                   pypi_0    pypi
dill                      0.3.7                    pypi_0    pypi
docker-pycreds            0.4.0                    pypi_0    pypi
docstring-parser          0.15                     pypi_0    pypi
einops                    0.7.0                    pypi_0    pypi
exceptiongroup            1.2.0                    pypi_0    pypi
fastapi                   0.109.0                  pypi_0    pypi
ffmpy                     0.3.1                    pypi_0    pypi
filelock                  3.13.1                   pypi_0    pypi
flash-attn                2.4.3.post1              pypi_0    pypi
fonttools                 4.47.2                   pypi_0    pypi
frozenlist                1.4.1                    pypi_0    pypi
fsspec                    2023.10.0                pypi_0    pypi
gitdb                     4.0.11                   pypi_0    pypi
gitpython                 3.1.41                   pypi_0    pypi
gradio                    3.50.2                   pypi_0    pypi
gradio-client             0.6.1                    pypi_0    pypi
h11                       0.14.0                   pypi_0    pypi
hjson                     3.1.0                    pypi_0    pypi
httpcore                  1.0.2                    pypi_0    pypi
httpx                     0.26.0                   pypi_0    pypi
huggingface-hub           0.20.3                   pypi_0    pypi
idna                      3.6                      pypi_0    pypi
importlib-resources       6.1.1                    pypi_0    pypi
jieba                     0.42.1                   pypi_0    pypi
jinja2                    3.1.3                    pypi_0    pypi
joblib                    1.3.2                    pypi_0    pypi
jsonschema                4.21.1                   pypi_0    pypi
jsonschema-specifications 2023.12.1                pypi_0    pypi
kiwisolver                1.4.5                    pypi_0    pypi
ld_impl_linux-64          2.38                 h1181459_1  
libffi                    3.4.4                h6a678d5_0  
libgcc-ng                 11.2.0               h1234567_1  
libgomp                   11.2.0               h1234567_1  
libstdcxx-ng              11.2.0               h1234567_1  
libuuid                   1.41.5               h5eee18b_0  
markdown-it-py            3.0.0                    pypi_0    pypi
markupsafe                2.1.4                    pypi_0    pypi
matplotlib                3.8.2                    pypi_0    pypi
mdurl                     0.1.2                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
multidict                 6.0.4                    pypi_0    pypi
multiprocess              0.70.15                  pypi_0    pypi
ncurses                   6.4                  h6a678d5_0  
networkx                  3.2.1                    pypi_0    pypi
ninja                     1.11.1.1                 pypi_0    pypi
nltk                      3.8.1                    pypi_0    pypi
numpy                     1.26.3                   pypi_0    pypi
nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
nvidia-cudnn-cu12         8.9.2.26                 pypi_0    pypi
nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
nvidia-nccl-cu12          2.18.1                   pypi_0    pypi
nvidia-nvjitlink-cu12     12.3.101                 pypi_0    pypi
nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
openssl                   3.0.12               h7f8727e_0  
orjson                    3.9.12                   pypi_0    pypi
packaging                 23.2                     pypi_0    pypi
pandas                    2.2.0                    pypi_0    pypi
peft                      0.7.1                    pypi_0    pypi
pillow                    10.2.0                   pypi_0    pypi
pip                       23.3.1          py310h06a4308_0  
protobuf                  4.25.2                   pypi_0    pypi
psutil                    5.9.8                    pypi_0    pypi
py-cpuinfo                9.0.0                    pypi_0    pypi
pyarrow                   15.0.0                   pypi_0    pypi
pyarrow-hotfix            0.6                      pypi_0    pypi
pydantic                  2.5.3                    pypi_0    pypi
pydantic-core             2.14.6                   pypi_0    pypi
pydub                     0.25.1                   pypi_0    pypi
pygments                  2.17.2                   pypi_0    pypi
pynvml                    11.5.0                   pypi_0    pypi
pyparsing                 3.1.1                    pypi_0    pypi
python                    3.10.13              h955ad1f_0  
python-dateutil           2.8.2                    pypi_0    pypi
python-multipart          0.0.6                    pypi_0    pypi
pytz                      2023.3.post1             pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
readline                  8.2                  h5eee18b_0  
referencing               0.32.1                   pypi_0    pypi
regex                     2023.12.25               pypi_0    pypi
requests                  2.31.0                   pypi_0    pypi
rich                      13.7.0                   pypi_0    pypi
rouge-chinese             1.0.3                    pypi_0    pypi
rpds-py                   0.17.1                   pypi_0    pypi
safetensors               0.4.1                    pypi_0    pypi
scipy                     1.12.0                   pypi_0    pypi
semantic-version          2.10.0                   pypi_0    pypi
sentencepiece             0.1.99                   pypi_0    pypi
sentry-sdk                1.39.2                   pypi_0    pypi
setproctitle              1.3.3                    pypi_0    pypi
setuptools                68.2.2          py310h06a4308_0  
shtab                     1.6.5                    pypi_0    pypi
six                       1.16.0                   pypi_0    pypi
smmap                     5.0.1                    pypi_0    pypi
sniffio                   1.3.0                    pypi_0    pypi
sqlite                    3.41.2               h5eee18b_0  
sse-starlette             1.8.2                    pypi_0    pypi
starlette                 0.35.1                   pypi_0    pypi
sympy                     1.12                     pypi_0    pypi
tiktoken                  0.5.2                    pypi_0    pypi
tk                        8.6.12               h1ccaba5_0  
tokenizers                0.15.0                   pypi_0    pypi
toolz                     0.12.0                   pypi_0    pypi
**torch                     2.1.2                   pypi_0    pypi**
tqdm                      4.66.1                   pypi_0    pypi
transformers              4.36.3.dev0              pypi_0    pypi
triton                    2.1.0                    pypi_0    pypi
trl                       0.7.10                   pypi_0    pypi
typing-extensions         4.9.0                    pypi_0    pypi
tyro                      0.6.6                    pypi_0    pypi
tzdata                    2023.4                   pypi_0    pypi
urllib3                   2.1.0                    pypi_0    pypi
uvicorn                   0.26.0                   pypi_0    pypi
wandb                     0.16.2                   pypi_0    pypi
websockets                11.0.3                   pypi_0    pypi
wheel                     0.41.2          py310h06a4308_0  
xformers                  0.0.23.post1             pypi_0    pypi
xxhash                    3.4.1                    pypi_0    pypi
xz                        5.4.5                h5eee18b_0  
yarl                      1.9.4                    pypi_0    pypi
zlib                      1.2.13               h5eee18b_0

**Comment by AGI-player** [comment]
> Hey would you mind providing more detail here, as well as a minimal reproducing script.

while, i just test with torch==2.1.1, it cannot reproduce the results either.

**Comment by drisspg** [comment]
What exactly is the issues? The images your provided aren't very clear to me

**Comment by AGI-player** [comment]
> What exactly is the issue? The images your provided aren't very clear to me

i run the train script with torch version==2.1.0, the first time train loss as follws
<img width="385" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/540dbe96-f6a7-4ac1-83bf-9c4b8a68e7c1">
the second time i run the script, just for repeating the training process, train loss at the same step as follows
<img width="396" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/7a9e9b8a-e7c4-43df-be11-467c78c4a446">
i only show parts of the train loss. In fact, in the whole training process they are always the same with each other as expected.

while, i run the train script with torch version==2.1.2, the first time train loss as follws
<img width="381" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/f3e06bf4-26f1-4ddd-b428-1da42d9cec73">
the second time, train loss at the same step as follows
<img width="382" alt="image" src="https://github.com/pytorch/pytorch/assets/99712469/76cea51a-dfb1-4642-aed4-48bc69ced211">
they are not always the same with each other in the whole training process.

**Comment by AGI-player** [comment]
> There are many things that could be leading to this; dataset iteration, deterministic ops, random seeds just to name a few.
> 
> Broadly this falls under the subject of reproducibility.
> 
> Here is a doc explaining the topic https://pytorch.org/docs/stable/notes/randomness.html
> 
> I don't think that what you posted classifies as an issue and I will be closing this for now.

of course, many things may lead to this result. But, in the compare experiments as above, the only difference is the torch version(2.1.0 vs 2.1.2), all the other parameters and conda environments are the same. So the difference(the training process can or cannot be reproduced) most likely is caused by torch's differnet version. 
whatever, i just found these results and show them here.
